flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogStorageLoader.java

```diff
@@ -28,16 +34,63 @@
-        this.pluginManager = pluginManager;
+    private static final Logger LOG = LoggerFactory.getLogger(StateChangelogStorageLoader.class);
+
+    /**
+     * Mapping of state changelog storage identifier to the corresponding storage factories,
+     * populated in {@link StateChangelogStorageLoader#initialize(PluginManager)}.
+     */
+    private static final HashMap<String, StateChangelogStorageFactory>
+            STATE_CHANGELOG_STORAGE_FACTORIES = new HashMap<>();
+
+    static {
+        // Guarantee to trigger once.
+        initialize(null);
     }
 
-    @SuppressWarnings({"rawtypes"})
-    public Iterator<StateChangelogStorage> load() {
-        return concat(
-                pluginManager.load(StateChangelogStorage.class),
-                ServiceLoader.load(StateChangelogStorage.class).iterator());
+    public static void initialize(PluginManager pluginManager) {
+        STATE_CHANGELOG_STORAGE_FACTORIES.clear();
+        Iterator<StateChangelogStorageFactory> iterator =
+                pluginManager == null
+                        ? ServiceLoader.load(StateChangelogStorageFactory.class).iterator()
+                        : concat(
+                                pluginManager.load(StateChangelogStorageFactory.class),
+                                ServiceLoader.load(StateChangelogStorageFactory.class).iterator());
+        iterator.forEachRemaining(
+                factory -> {
+                    String identifier = factory.getIdentifier().toLowerCase();
+                    StateChangelogStorageFactory prev =
+                            STATE_CHANGELOG_STORAGE_FACTORIES.get(identifier);
+                    if (prev == null) {
+                        STATE_CHANGELOG_STORAGE_FACTORIES.put(identifier, factory);
+                    } else {
+                        LOG.warn(
+                                "StateChangelogStorageLoader found duplicated factory,"
+                                        + " using {} instead of {} for name {}.",
+                                prev.getClass().getName(),
+                                factory.getClass().getName(),
+                                identifier);
+                    }
+                });
+        LOG.info(
+                "StateChangelogStorageLoader initialized with shortcut names {{}}.",
+                String.join(",", STATE_CHANGELOG_STORAGE_FACTORIES.keySet()));
+    }
+
+    public static StateChangelogStorage<?> load(Configuration configuration) {
+        final String identifier =
+                configuration
+                        .getString(CheckpointingOptions.STATE_CHANGE_LOG_STORAGE)
+                        .toLowerCase();
+
+        StateChangelogStorageFactory factory = STATE_CHANGELOG_STORAGE_FACTORIES.get(identifier);
+        if (factory == null) {
+            LOG.warn("Cannot find a factory for changelog storage with name '{}'.", identifier);
+            return null;
```

Comment:
1. If the storage isn't needed then `STATE_CHANGE_LOG_STORAGE` is empty, right? Why don't we return `Optional.empty` in that case? (`null` isn't very obvious, and should be marked `Nullable` if we keep it)
2. If an invalid identifier is provided, why don't we signal it explicitly (by throwing an exception for example)? I think returning `null` (or empty Optional) can hide misconfiguration issues.

Reply:
1. Currently there is a default value for ```STATE_CHANGE_LOG_STORAGE ``` so we won't run into this case. However, a ```Nullable``` mark here makes sense. Will do.
2. I'd rather keep it return null here. User may disable changelog and config a wrong value for  ```STATE_CHANGE_LOG_STORAGE ``` (by some copy&paste operation or missing jars for factory). Since they have disable the changelog, the changelog-related configuration should better not make the job FAIL. WDYT?

Reply:
Fair enough, removing jars and disabling changelog while keeping some value `state.backend.changelog.storage` seems a valid operation.